plugins/mips/mips_logic.ml

```diff
@@ -0,0 +1,182 @@
+let mand cpu ops =
+  let rd = unsigned cpu.reg ops.(0) in
+  let rs = unsigned cpu.reg ops.(1) in
+  let rt = unsigned cpu.reg ops.(2) in
+  RTL.[
+    rd := rs land rt;
+  ]
+
+(* ANDI rd, rs, imm
+ * And Immediate, MIPS32
+ * Page 48 *)
+let mandi cpu ops =
+  let rd = unsigned cpu.reg ops.(0) in
+  let rs = unsigned cpu.reg ops.(1) in
+  let im = unsigned imm ops.(2) in
+  RTL.[
+    rd := rs land im;
+  ]
+
+(* TODO: EXT rd, rs, p, s *)
+(* TODO: INS rd, rs, p, s *)
+
+(* NOP
+ * Just NOP, here and everywhere
+ * Page 364 *)
+let nop cpu ops = []
+
+(* NOR rd, rs, rt
+ * Not Or, MIPS32
+ * Page 365 *)
+let nor cpu ops =
+  let rd = unsigned cpu.reg ops.(0) in
+  let rs = unsigned cpu.reg ops.(1) in
+  let rt = unsigned cpu.reg ops.(2) in
+  RTL.[
+    rd := lnot (rs lor rt);
+  ]
+
+(* NOT rd, rs
+ * Not in the manual *)
+let mnot cpu ops =
+  let rd = unsigned cpu.reg ops.(0) in
+  let rs = unsigned cpu.reg ops.(1) in
+  RTL.[
+    rd := lnot rs;
+  ]
+
+(* OR rd, rs, rt
+ * Or, MIPS32
+ * Page 366 *)
+let mor cpu ops =
+  let rd = unsigned cpu.reg ops.(0) in
+  let rs = unsigned cpu.reg ops.(1) in
+  let rt = unsigned cpu.reg ops.(2) in
+  RTL.[
+    rd := rs lor rt;
+  ]
+
+(* ORI rd, rs, imm
+ * Or Immediate, MIPS32
+ * Page 367 *)
+let mori cpu ops =
+  let rd = unsigned cpu.reg ops.(0) in
+  let rs = unsigned cpu.reg ops.(1) in
+  let im = unsigned imm ops.(2) in
+  RTL.[
+    rd := rs lor im;
+  ]
+
+(* WSBH rd, rs
+ * Word Swap Bytes Within Halfwords, MIPS32 Release 2
+ * Page 509 *)
+let wsbh cpu ops =
+  let rd = unsigned cpu.reg ops.(0) in
+  let rs = unsigned cpu.reg ops.(1) in
+  RTL.[
+    rd := nth byte rs 3 ^ nth byte rs 4 ^
+      nth byte rs 1 ^ nth byte rs 2;
+  ]
+
+(* DSBH rd, rs
+ * Doubleord Swap Bytes Within Halfwords, MIPS64 Release 2
+ * Page 196 *)
+let dsbh cpu ops =
+  let rd = unsigned cpu.reg ops.(0) in
+  let rs = unsigned cpu.reg ops.(1) in
+  RTL.[
+    rd := nth byte rs 7 ^ nth byte rs 8 ^
+      nth byte rs 5 ^ nth byte rs 6 ^
+      nth byte rs 3 ^ nth byte rs 4 ^
+      nth byte rs 1 ^ nth byte rs 2;
+  ]
+
+(* DSHD rd, rs
+ * Doubleord Swap Halfword Within Doublewords, MIPS64 Release 2
+ * Page 509 *)
+let dshd cpu ops =
+  let rd = unsigned cpu.reg ops.(0) in
+  let rs = unsigned cpu.reg ops.(1) in
+  RTL.[
+    rd := nth halfword rs 1 ^ nth halfword rs 2 ^
+      nth halfword rs 3 ^ nth halfword rs 4;
+  ]
+
+(* XOR rd, rs, rt
+ * Exclusive Or, MIPS32
+ * Page 509 *)
+let mxor cpu ops =
+  let rd = unsigned cpu.reg ops.(0) in
+  let rs = unsigned cpu.reg ops.(1) in
+  let rt = unsigned cpu.reg ops.(2) in
+  RTL.[
+    rd := rs lxor rt;
+  ]
+
+(* XORI rd, rs, imm
+ * Exclusive Or Immediate, MIPS32
+ * Page 511 *)
+let mxori cpu ops =
+  let rd = unsigned cpu.reg ops.(0) in
+  let rs = unsigned cpu.reg ops.(1) in
+  let im = unsigned imm ops.(2) in
+  RTL.[
+    rd := rs lxor im;
+  ]
+
+(* BITSWAP rd, rt
+ * Swaps (reverses) bits in each byte, MIPS32 Release 6
+ * Page 95 *)
+let bitswap cpu ops =
```

Comment:
We shall try to make this operation without a loop, e.g. as [here][1]. Adding a loop here will increase the complexity of a binary and hamper analysis, though the operation by itself has a very limited side effect, it is really a pain to verify this formally. 


[1]: https://stackoverflow.com/questions/21511533/reverse-integer-bitwise-without-using-loop